saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationRequestFactory.java

```diff
@@ -27,15 +34,40 @@
  */
 public interface Saml2AuthenticationRequestFactory {
 	/**
-	 * Creates an authentication request from the Service Provider, sp,
-	 * to the Identity Provider, idp.
+	 * Creates an authentication request from the Service Provider, sp, to the Identity Provider, idp.
 	 * The authentication result is an XML string that may be signed, encrypted, both or neither.
+	 * This method only returns the {@code SAMLRequest} string for the request, and for a complete
+	 * set of data parameters please use {@link #createAuthenticationRequestParameters(Saml2AuthenticationRequest)}
 	 *
 	 * @param request - information about the identity provider, the recipient of this authentication request and
 	 *                accompanying data
 	 * @return XML data in the format of a String. This data may be signed, encrypted, both signed and encrypted or
 	 * neither signed and encrypted
 	 * @throws Saml2Exception when a SAML library exception occurs
+	 * @deprecated Use {@link #createAuthenticationRequestParameters(Saml2AuthenticationRequest)} to retrieve all parameters
 	 */
 	String createAuthenticationRequest(Saml2AuthenticationRequest request);
+
+	/**
+	 * Creates all the necessary AuthNRequest parameters for a POST or REDIRECT binding.
+	 * If the {@link Saml2AuthenticationRequest} doesn't contain any {@link Saml2X509CredentialType#SIGNING} credentials
+	 * the result will not contain any signatures.
+	 * If {@link Saml2AuthenticationRequest#getBinding()} returns true, the data set will be signed and encoded for
+	 * REDIRECT binding and DEFLATE encoding.
+	 * If {@link Saml2AuthenticationRequest#getBinding()} returns false, the data set will be signed and encoded for
+	 * POST binding and if applicable signed with XML signatures.
+	 *
+	 * @param request - information about the identity provider, the recipient of this authentication request and
+	 *                accompanying data
+	 * @return a new map with existing query parameters and the necessary parameters to make the AuthNRequest
+	 * over a POST or REDIRECT binding.
+	 * @throws Saml2Exception when a SAML library exception occurs
+	 */
+	default Saml2AuthenticationRequest createAuthenticationRequestParameters(Saml2AuthenticationRequest request) {
```

Comment:
Also, I think that a contract of `(Saml2AuthenticationRequest.Builder, RelyingPartyRegistration) -> Saml2AuthenticationRequest` is worth considering. I think this communicates more clearly that what is being passed in is a prototype for what will ultimately be created.

Reply:
>Since this is creating a Saml2AuthenticationRequest, should it now be called createAuthenticationRequest?

I wish, but Java doesn't allow you to overload when only the response type changes.

> I think that a contract of (Saml2AuthenticationRequest.Builder, RelyingPartyRegistration) -> Saml2AuthenticationRequest is worth considering

I'm not following. Where would this contact be? Passing a builder into an interface?

Reply:
Addressed in: ae238b1ee93233e6a5413faa4408376c0d711273

This also opens up the option to separate input from output by creating `Saml2AuthenticationRequestToken` (name may vary)